File: uniffi/src/bindings/swift/templates/ErrorTemplate.swift

public enum {{e.name()}}: LocalizedError {
    {% for value in e.values() %}
    case {{value}}(message: String)
    {% endfor %}


    /// Our implementation of the localizedError protocol
    public var errorDescription: String? {
        switch self {
        {% for value in e.values() %}
        case let .{{value}}(message):
            return "{{e.name()}}.{{value}}: \(message)"
        {% endfor %}
        }
    }

    // The name is attempting to indicate that we free message if it
    // existed, and that it's a very bad idea to touch it after you call this
    // function
    static func fromConsuming(_ rustError: Native{{e.name()}}) throws -> {{e.name()}}? {
        let message = rustError.message
        switch rustError.code {
            case {{e.name()}}_NoError:
                return nil
            {% for value in e.values() %}
            case {{e.name()}}_{{value}}:
                return .{{value}}(message: try String.fromFFIValue(message!))
            {% endfor %}
            default:
                return nil
        }
    }

    @discardableResult
    public static func unwrap<T>(_ callback: (UnsafeMutablePointer<Native{{e.name()}}>) throws -> T?) throws -> T {
        guard let result = try tryUnwrap(callback) else {
            throw {{e.name()}}ResultError.empty
        }
        return result
    }

    @discardableResult
    public static func tryUnwrap<T>(_ callback: (UnsafeMutablePointer<Native{{e.name()}}>) throws -> T?) throws -> T? {
        var err = Native{{e.name()}}(code: {{e.name()}}_NoError, message: nil)
        let returnedVal = try callback(&err)
        if let retErr = try {{e.name()}}.fromConsuming(err) {
            throw retErr
        }
        return returnedVal
    }
}

internal func rustCallWith{{e.name()}}<T>(_ cb: (UnsafeMutablePointer<Native{{e.name()}}>) throws -> T?) throws -> T {
    return try {{e.name()}}.unwrap { err in
        return try cb(err)
    }
}

internal func nullableRustCallWith{{e.name()}}<T>(_ cb: (UnsafeMutablePointer<Native{{e.name()}}>) throws -> T?) throws -> T? {
    return try {{e.name()}}.tryUnwrap { err in
        return try cb(err)
    }
}

public enum {{e.name()}}ResultError: Error {
    case empty
}

File: uniffi/src/bindings/kotlin/templates/ErrorTemplate.kt
@Structure.FieldOrder("code", "message")
internal open class {{e.name()}} : Structure() {

    class ByReference : {{e.name()}}(), Structure.ByReference

    @JvmField var code: Int = 0
    @JvmField var message: Pointer? = null

    /**
     * Does this represent success?
     */
    fun isSuccess(): Boolean {
        return code == 0
    }

    /**
     * Does this represent failure?
     */
    fun isFailure(): Boolean {
        return code != 0
    }

    @Suppress("ReturnCount", "TooGenericExceptionThrown")
    fun intoException(): {{e.name()}}Exception {
        if (!isFailure()) {
            // It's probably a bad idea to throw here! We're probably leaking something if this is
            // ever hit! (But we shouldn't ever hit it?)
            throw RuntimeException("[Bug] intoException called on non-failure!")
        }
        val message = this.consumeErrorMessage()
        when (code) {
            {% for value in e.values() -%}
            {{loop.index}} -> return {{e.name()}}Exception.{{value}}(message)
            {% endfor -%}
            else -> throw RuntimeException("Invalid error recieved...")
        }
    }

    /**
     * Get and consume the error message, or null if there is none.
     */
    @Synchronized
    fun consumeErrorMessage(): String {
        val result = this.getMessage()
        if (this.message != null) {
            this.ensureConsumed()
        }
        if (result == null) {
            throw NullPointerException("consumeErrorMessage called with null message!")
        }
        return result
    }

    @Synchronized
    fun ensureConsumed() {
        if (this.message != null) {
            _UniFFILib.INSTANCE.{{ci.namespace()}}_string_free(this.message!!)
            this.message = null
        }
    }

    /**
     * Get the error message or null if there is none.
     */
    fun getMessage(): String? {
        return this.message?.getString(0, "utf8")
    }
}

open class {{e.name()}}Exception(message: String) : Exception(message) {
    {% for value in e.values() -%}
    class {{value}}(msg: String) : {{e.name()}}Exception(msg)
    {% endfor %}
}

// Helpers for calling Rust with errors:
// In practice we usually need to be synchronized to call this safely, so it doesn't
// synchronize itself
private inline fun <U> nullableRustCallWith{{e.name()}}(callback: ({{e.name()}}.ByReference) -> U?): U? {
    val e = {{e.name()}}.ByReference()
    try {
        val ret = callback(e)
        if (e.isFailure()) {
            throw e.intoException()
        }
        return ret
    } finally {
        // This only matters if `callback` throws (or does a non-local return, which
        // we currently don't do)
        e.ensureConsumed()
    }
}

private inline fun <U> rustCallWith{{e.name()}}(callback: ({{e.name()}}.ByReference) -> U?): U {
    return nullableRustCallWith{{e.name()}}(callback)!!
}
